### PR TITLE
feat(testing): Fluent api to test source code transformations.

### LINKED ISSUE #499: feature: one liner test infrastructure
(suggested by @rouvoy)
Spec:

    assertThat('Foo.java').transformWith(MyProcessor.class).is('FooTransform.java')

What do you think?

### PR DESCRIPTION
**Pull request by GerardPaligot**
Closes #499

### CODE REVIEW

**Comment by GerardPaligot** [comment]
@monperrus What do you think about this PR?

File: src/main/java/spoon/testing/utils/ModelUtils.java [hunk 1]
```diff
@@ -74,7 +101,7 @@ public static void canBeBuilt(File outputDirectoryFile, int complianceLevel, boo
 		try {
 			compiler.build();
 		} catch (Exception e) {
-			throw new AssertionError("Can't compile " + outputDirectoryFile.getName(), e);
+			throw new AssertionError("Can't compile " + outputDirectoryFile.getName());
```
Comment (GerardPaligot) [inline src/main/java/spoon/testing/utils/ModelUtils.java]: display message of `e`.

File: src/test/java/spoon/testing/Config.java [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package spoon.testing;
+
+public class Config {
+	public static final String PATH = "./src/test/java/spoon/testing/testclasses/";
```
Comment (GerardPaligot) [inline src/test/java/spoon/testing/Config.java]: copy this field in all test classes.

File: src/test/java/spoon/testing/FactoryAssertTest.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import spoon.testing.testclasses.Bar;
+import spoon.testing.testclasses.Foo;
+
+import java.io.File;
+
+import static spoon.testing.Assert.assertThat;
+import static spoon.testing.Config.PATH;
+import static spoon.testing.utils.ModelUtils.build;
+
+public class FactoryAssertTest {
```
Comment (GerardPaligot) [inline src/test/java/spoon/testing/FactoryAssertTest.java]: Remove equals factory.

File: src/main/java/spoon/testing/AbstractAssert.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+	 * @return {@code this} assertion object.
+	 */
+	protected T isEqualTo(Factory expected, Factory actual) {
+		process(actual, processors);
+		final List<CtType<?>> allActual = actual.Type().getAll();
+		final List<CtType<?>> allExpected = expected.Type().getAll();
+		for (int i = 0; i < allActual.size(); i++) {
+			final CtType<?> currentActual = allActual.get(i);
+			final CtType<?> currentExpected = allExpected.get(i);
+			if (!currentActual.equals(currentExpected)) {
```
Comment (GerardPaligot) [inline src/main/java/spoon/testing/AbstractAssert.java]: Equals on toString and not signature.

**Comment by GerardPaligot** [comment]
Adds documentation about testing "module".

**Comment by GerardPaligot** [comment]
@monperrus What do you think about this PR now?

**Comment by GerardPaligot** [comment]
@monperrus Ok for you https://github.com/GerardPaligot/spoon/commit/43f8559d1ce77215f1d1f7c3b28baef088b37273 ?

**Comment by GerardPaligot** [comment]
A rebase will be necessary after the merge of this PR: https://github.com/INRIA/spoon/pull/515

**Comment by monperrus** [comment]
go for the rebase!

**Pull request merged**